File: Sources/UsageWidget/Views/ClaudeMascot.swift
import SwiftUI

/// A tiny pixel-art Claude mascot that idles in place — hopping, blinking, glancing around, and
/// shuffling its feet — a nod to the animated companion in Claude Code Desktop's chat input.
///
/// Pure SwiftUI `Canvas` + `TimelineView(.animation)`, so it only ticks while on screen. In a
/// `MenuBarExtra` the status-bar *label* never gets a display link (so animation there is dead),
/// but the popover is a real window that does — hence this lives in the popover header. When the
/// popover closes the view is torn down and the timeline stops, so it costs nothing at rest.
///
/// Everything is drawn from a small sprite grid. The body is static; eyes and legs are painted
/// procedurally on top so they can blink / dart / march independently of the body.
struct ClaudeMascotView: View {
    /// Size of one sprite pixel, in points. The mascot spans `gridW × (gridH + 1)` of these.
    /// 1.5pt → 3 device pixels on a 2× display, keeping the art crisp (everything snaps to cells).
    var pixel: CGFloat = 1.5

    // Sprite body: ' ' transparent, 'B' body. 10 rows; legs occupy the two rows below it.
    private static let body: [String] = [
        "       BB       ",
        "    BBBBBBBB    ",
        "   BBBBBBBBBB   ",
        "  BBBBBBBBBBBB  ",
        " BBBBBBBBBBBBBB ",
        " BBBBBBBBBBBBBB ",
        " BBBBBBBBBBBBBB ",
        " BBBBBBBBBBBBBB ",
        "  BBBBBBBBBBBB  ",
        "   BBBBBBBBBB   ",
    ]
    private static let gridW = 16
    private static let gridH = 12   // 10 body rows + 2 leg rows

    // Palette — Claude brand terracotta over the popover's dark material.
    private static let bodyColor = Color(hex: "#D97757")
    private static let legColor  = Color(hex: "#B05A3C")
    private static let eyeWhite  = Color(hex: "#F8FAFC")
    private static let pupil     = Color(hex: "#1B1B1F")

    // Top-left cell of each 2×2 eye; 1px-wide leg columns.
    private static let eyes = [(x: 3, y: 3), (x: 6, y: 3)]
    private static let legX = [3, 6, 9, 12]

    var body: some View {
        TimelineView(.animation) { timeline in
            Canvas { ctx, _ in
                draw(ctx, t: timeline.date.timeIntervalSinceReferenceDate)
            }
            .frame(width: CGFloat(Self.gridW) * pixel,
                   height: CGFloat(Self.gridH + 1) * pixel)   // +1 row of headroom for the hop
        }
        .accessibilityHidden(true)
    }

    private func px(_ ctx: GraphicsContext, _ x: Int, _ y: Int, _ color: Color, _ dy: CGFloat) {
        let r = CGRect(x: CGFloat(x) * pixel, y: CGFloat(y) * pixel + dy, width: pixel, height: pixel)
        ctx.fill(Path(r), with: .color(color))
    }

    private func draw(_ ctx: GraphicsContext, t: TimeInterval) {
        // Idle hop: rests 1px low, rises flush to the top on a ~0.6s cadence.
        let bobY: CGFloat = sin(t * 5) > 0 ? 0 : pixel

        // Static body.
        for (row, line) in Self.body.enumerated() {
            for (col, ch) in line.enumerated() where ch == "B" {
                px(ctx, col, row, Self.bodyColor, bobY)
            }
        }
        // A darker snout pixel just below the eyes.
        px(ctx, 5, 6, Self.legColor, bobY)

        // Legs: each foot plants or tucks up, alternating sets for a march-in-place shuffle.
        let step = Int(t * 2.5) % 2
        for (i, lx) in Self.legX.enumerated() {
            px(ctx, lx, Self.gridH - 2, Self.legColor, bobY)                       // upper leg, always
            if (i + step) % 2 != 0 { px(ctx, lx, Self.gridH - 1, Self.legColor, bobY) }  // foot when planted
        }

        // Eyes: a quick blink every ~3.2s; otherwise 2×2 whites with a pupil that darts the corners.
        let blink = t.truncatingRemainder(dividingBy: 3.2) < 0.14
        let (dx, dy) = [(0, 1), (1, 1), (1, 0), (0, 0)][Int(t / 1.6) % 4]
        for e in Self.eyes {
            if blink {
                px(ctx, e.x, e.y + 1, Self.legColor, bobY)            // closed lid
                px(ctx, e.x + 1, e.y + 1, Self.legColor, bobY)
            } else {
                px(ctx, e.x, e.y, Self.eyeWhite, bobY)
                px(ctx, e.x + 1, e.y, Self.eyeWhite, bobY)
                px(ctx, e.x, e.y + 1, Self.eyeWhite, bobY)
                px(ctx, e.x + 1, e.y + 1, Self.eyeWhite, bobY)
                px(ctx, e.x + dx, e.y + dy, Self.pupil, bobY)         // darting pupil
            }
        }
    }
}

File: windows/src/App/Views/ClaudeMascot.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Media;
using Avalonia.Threading;

namespace LLMUsageWidget.App.Views;

/// <summary>A tiny pixel-art Claude mascot that idles in the popover header — hopping, blinking,
/// glancing around, and shuffling its feet. Ported from the macOS app's ClaudeMascotView (same
/// sprite grid + timings). Drawn directly to the <see cref="DrawingContext"/>; a 30 fps timer ticks
/// only while attached, so it costs nothing once the popover closes.</summary>
public sealed class ClaudeMascot : Control
{
    private const double Pixel = 1.5;   // size of one sprite cell → 16×13 cells ≈ 24×19.5 px
    private const int GridW = 16;
    private const int GridH = 12;       // 10 body rows + 2 leg rows

    // Sprite body: ' ' transparent, 'B' body. Legs occupy the two rows below it.
    private static readonly string[] Body =
    {
        "       BB       ",
        "    BBBBBBBB    ",
        "   BBBBBBBBBB   ",
        "  BBBBBBBBBBBB  ",
        " BBBBBBBBBBBBBB ",
        " BBBBBBBBBBBBBB ",
        " BBBBBBBBBBBBBB ",
        " BBBBBBBBBBBBBB ",
        "  BBBBBBBBBBBB  ",
        "   BBBBBBBBBB   ",
    };

    private static readonly IBrush BodyColor = new SolidColorBrush(Color.Parse("#D97757"));
    private static readonly IBrush LegColor = new SolidColorBrush(Color.Parse("#B05A3C"));
    private static readonly IBrush EyeWhite = new SolidColorBrush(Color.Parse("#F8FAFC"));
    private static readonly IBrush Pupil = new SolidColorBrush(Color.Parse("#1B1B1F"));

    private static readonly (int X, int Y)[] Eyes = { (3, 3), (6, 3) };
    private static readonly int[] LegX = { 3, 6, 9, 12 };
    private static readonly (int Dx, int Dy)[] Gaze = { (0, 1), (1, 1), (1, 0), (0, 0) };

    private readonly DispatcherTimer _timer;
    private double _t = 0.3;   // a calm opening frame (body up, eyes open) for the still snapshot

    public ClaudeMascot()
    {
        Width = GridW * Pixel;
        Height = (GridH + 1) * Pixel;   // +1 row of headroom for the hop
        _timer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(1000.0 / 30) };
        _timer.Tick += (_, _) => { _t += 1.0 / 30; InvalidateVisual(); };
    }

    protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
    {
        base.OnAttachedToVisualTree(e);
        _timer.Start();
    }

    protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e)
    {
        base.OnDetachedFromVisualTree(e);
        _timer.Stop();
    }

    private static void Px(DrawingContext ctx, int x, int y, IBrush color, double dy) =>
        ctx.FillRectangle(color, new Rect(x * Pixel, y * Pixel + dy, Pixel, Pixel));

    public override void Render(DrawingContext ctx)
    {
        double t = _t;
        double bob = Math.Sin(t * 5) > 0 ? 0 : Pixel;   // idle hop: rests 1px low, rises flush

        for (int row = 0; row < Body.Length; row++)
            for (int col = 0; col < Body[row].Length; col++)
                if (Body[row][col] == 'B')
                    Px(ctx, col, row, BodyColor, bob);

        Px(ctx, 5, 6, LegColor, bob);   // darker snout pixel below the eyes

        int step = (int)(t * 2.5) % 2;
        for (int i = 0; i < LegX.Length; i++)
        {
            Px(ctx, LegX[i], GridH - 2, LegColor, bob);                          // upper leg, always
            if ((i + step) % 2 != 0) Px(ctx, LegX[i], GridH - 1, LegColor, bob); // foot when planted
        }

        bool blink = t % 3.2 < 0.14;
        var (dx, dy) = Gaze[(int)(t / 1.6) % 4];
        foreach (var (ex, ey) in Eyes)
        {
            if (blink)
            {
                Px(ctx, ex, ey + 1, LegColor, bob);
                Px(ctx, ex + 1, ey + 1, LegColor, bob);
            }
            else
            {
                Px(ctx, ex, ey, EyeWhite, bob);
                Px(ctx, ex + 1, ey, EyeWhite, bob);
                Px(ctx, ex, ey + 1, EyeWhite, bob);
                Px(ctx, ex + 1, ey + 1, EyeWhite, bob);
                Px(ctx, ex + dx, ey + dy, Pupil, bob);   // darting pupil
            }
        }
    }
}
